pkg/agent/secondarynetwork/podwatch/controller.go

```diff
@@ -0,0 +1,402 @@
+	var err error
+	var ipamerr error
+	var ipamResult *current.Result
+	var cmdArgs *invoke.Args
+
+	// Assigne default interface name, If secondary interface name was not provided in Pod annotation.
+	if len(netinfo.InterfaceName) == 0 {
+		netinfo.InterfaceName = defaultSecondaryInterfaceName
+	}
+
+	for _, containerInfo := range podCNIInfo {
+		// CNI Server CmdDel handle will update CNI cache entry's PodStatus as "podDelComplete" upon successful clean-up of primary network config.
+		// Here (configureSecondaryInterface) we check for the PodStatus entry and do not process cache (as part of the current Pod Add event handle sequence), if its
+		// already marked as podDelComplete by CNI Server. CNI cache with PodStatus set to "podDelComplete" will be cleaned up as part of handleRemovePod (Pod Del event handle).
+		// NOTE: This additional status check was handled to avoid duplicate cache processing for the same PodName and PodNamespace.
+		// PodSecondaryIface shows the status of secondary network interface configuration. If already configured, avoid processing it again (As per the current implementation,
+		// we do not have full support for Pod update event handle (i'e, Pod annotation changes etc.,).
+		// TBD: Impelment Linux interface configuration API to look for secondary network interface in container scope and replace PodSecondaryIface check with the new API (Ideal
+		// way to handle).
+		if containerInfo.PodCNIStatus != podDelComplete && containerInfo.PodSecondaryIface == false {
+			// Update Virtual Network configuration (retrieved from network-attach-definition) at CNI cache. This is required to release IP address at whereabouts DB.
+			containerInfo.NetworkInfo = cniconfig
+			containerInfo.ContainerIfDev = netinfo.InterfaceName
+			pc.podCache.UpdateCNIConfigInfo(containerInfo)
+			if netinfo.InterfaceType == sriovInterfaceType {
+				cmdArgs = &invoke.Args{Command: string("ADD"), ContainerID: containerInfo.ContainerID,
+					NetNS: containerInfo.ContainerNetNS, IfName: netinfo.InterfaceName,
+					Path: cniPath}
+
+				ipamResult, err = ipam.GetIPAMSubnetAddress(cniconfig, cmdArgs)
+				if err != nil {
+					return errors.New("secondary network IPAM failed")
+				}
+
+				result := &current.Result{CNIVersion: containerInfo.CNIVersion}
+				result.IPs = ipamResult.IPs
+				result.Routes = ipamResult.Routes
+				// Set result.Interface to container interface.
+				for _, ip := range result.IPs {
+					ip.Interface = current.Int(1)
+				}
+				// Configure SRIOV as a secondary network interface
+				if err = pc.configureSriovAsSecondaryInterface(pod, netinfo, containerInfo, result); err != nil {
+					// SRIOV interface creation failed. Free allocated IP address
+					if ipamerr = ipam.DelIPAMSubnetAddress(cniconfig, cmdArgs); ipamerr != nil {
+						klog.ErrorS(err, "IPAM de-allocation failed: ", ipamerr)
```

Comment:
Can we add a comment saying if de-allocation fails, the IP might never be released?